StonedMC, the bastard love child of GoonPS and CarnMC

MrsTonedOne opened the pull request:
Basically, they key difference between StonedMC and CarnMC is that when multiple subsystems want to run at the same game tick, we divvy up the tick between the subsystems, rather then allow one subsystem to hog it all.  
The key difference between StonedMC and GoonPS is that we allow the subsystems to tell us how to divvy up the tick using flags and priority weighting.  
The new SS_ flags allows us to select behaviors that used to be piggybacked as side effects of dynamic wait or default but sometimes unneeded behavior.  
Dynamic wait is 100% gone, lower priority and SS_BACKGROUND are better more refined ways of doing this when combined with MC_TICK_CHECK  

I have by design never looked at the inners of goonPS, so this is all original code but I know goonPS uses two loops because of comments by goon devs on reddit threads, that design didn't make sense before, but when I can tell a SS how much of a game tick it is allowed to have, knowing how many need to run this tick is helpful. I also know a bit about how it works from piecing together comments in #vgstation.

Detailed list of changes:
- **Subsystems now have flags**, allowing fine grain control over things like rather or not it processes, initializes, rather it's wait is how long between runs (post run timing) or how long between starts, and rather or not late fires should cause the next fire to be earlier.
- **Mc now has two loops** One loop handles queuing logic, one loop handles running logic.
- **MC now splits up tick allotment rather than first come first serve** Subsystems can even request a bigger share using higher priorities. (It will even resume subsystems it paused if other subsystems hadn't used as much as it predicted they might need and there is spare time at the tail of a game tick)
- **Default tickrate is now 20** This is related enough to the MC and it's a change that's really long since over due

All code oddities are most likely to be necessities to lower overhead on the mc since it runs every tick

todo:
- [x] More testing.  
- [x] Failsafe testing.  
- [x] MC loop runtime detection and restart testing.  
- [x] Figure out why restarting the MC can break things randomly. (I have a few ideas)  
- [x] ~~Make timer faster and more like a mini mc in how it queues so it can take proper use of its new 1tick accuracy~~ Deferred (I'll do it in another pr)
- [x] Move some things over to init only subsystems (world/New() stuff, mc/Setup() bloat)  
- [x] ~~Move some things over to fire only subsystems (orbit, throw_at, space drift, projectiles)~~ Deferred (I'll do it in another pr)
- [x] Look at making air and lighting start up quicker
- [x] Remove the debugging lines.

fixes #17803 (mc now is the last thing to run in a tick, not the first, so other sleeps like progress bars don't overrun and get delayed by atmos)
fixes #18467 icon smoothing is fired much more often and throttled better

MrsTonedOne committed:
StonedMC, the bastard love child of GoonPS and CarnMC
Basically, they key difference between CarnMC is that when multiple ticks want to run at the same byond tick, we divvy up the tick between the subsystems, rather then allow one subsystem to hog it all.
The key difference between GoonPS is that we allow the subsystems to tell us how to divvy up the tick using flags and priority.
The new SS flags allows us to select behaviors that used to be piggy backed as side effects of dynamic wait.
dynamic wait is 100% gone, lower priority and SS_BACKGROUND are better more refined ways of doing this when combined with MC_TICK_CHECK

I have by design never looked at the inners of goonPS, but i know it uses two loops because of comments by goon devs on reddit threads, that design didn't make sense before, but when I can tell a SS how much of a byond tick it is allowed to have, knowing how many need to run this tick is helpful
All code oddities are most likely to be necessities to lower overhead on the mc since it runs every tick
todo:
Move some things over to init only subsystems (world/New() stuff, mc/Setup() bloat)
Move some things over to fire only subsystems (orbit, throw_at, space drift, projectiles)
Look at making air and lighting start up quicker
More testing.
Failsafe testing.
MC loop runtime detection and restart testing.
Figure out why restarting the MC can break things randomly. (I have a few ideas)

MrStonedOne commented on code/controllers/subsystem/fastprocess.dm:
```diff
@@ -2,13 +2,9 @@ var/datum/subsystem/fastprocess/SSfastprocess
-	dwait_delta = 6
+	priority = 25
+	flags = SS_BACKGROUND|SS_POST_FIRE_TIMING|SS_NO_INIT
+	wait = 4
```
todo, tweak and test these numbers.

Iamgoofball replied:
to clarify, fastprocess is still running as fast as humanely possible right?

MrStonedOne replied:
It is now... I bumped that down to 2ds.

It COULD go faster, but i don't want it eating up too much of the ticks.

MrStonedOne commented on code/controllers/subsystem.dm:
```diff
@@ -67,12 +67,8 @@
-
 	if(can_fire)
-		msg = "[round(cost,0.01)]ds|[round(tick_usage,1)]%\t[dwait][msg]"
+		msg = "[round(cost*ticks,1)]ms|[round(tick_usage,1)]%|[round(ticks,0.1)]\t[msg]"
```
That there is some good shit.

Millisecond time tracking for the mc.

PKPenguin321 commented:
I'm guessing this stomps on lag's already decomposing corpse even further?

Iamgoofball commented:
more like StonedAF, the MrStonedOne Action Filer

Razharas commented on code/controllers/master.dm:
```diff
@@ -120,129 +148,270 @@ var/global/datum/controller/master/Master = new()
+			continue //already firing
 		timer += world.tick_lag
 		SS.can_fire = 1
 		SS.next_fire = timer + rand(0, SS.wait) // Stagger subsystems.
```
Now that you have SS_KEEP_TIMING flag maybe its better to specifically spread those that use it evenly

MrStonedOne replied:
na, no matter what we benefit from them being evened out a bit,

Fox-McCloud commented on code/controllers/subsystem.dm:
```diff
@@ -2,33 +2,34 @@
 	// Metadata; you should define these.
-	var/name				//name of the subsystem
-	var/priority = 0		//priority affects order of initialization. Higher priorities are initialized first, lower priorities later. Can be decimal and negative values.
+	var/name = "fire coderbus" //name of the subsystem
```
;)

MrStonedOne replied:
ya, for a while, there was a subsystem with no name, and it looked stupid on the stat panel, so i added this so it looks less stupid next time somebody forgets to name a subsystem

pudl commented:
Hail mso

MrsTonedOne committed:
Default fps 20, debugging lines, fixes, subsystems made faster.

MrStonedOne commented on code/controllers/master.dm:
```diff
@@ -120,129 +148,301 @@ var/global/datum/controller/master/Master = new()
+		var/datum/subsystem/SS = thing
+		if (round_started || SS.flags & SS_FIRE_IN_LOBBY)
+			timer += world.tick_lag
+			SS.next_fire = timer
```
todo: merge this with the above loop.

MrStonedOne commented on code/__HELPERS/_logging.dm:
```diff
@@ -13,6 +13,9 @@
 #endif
+/proc/debug_world(msg)
+	if(Debug2)
+		world << "DEBUG: [msg]"
```
Why we didn't have this proc, i don't fucking know.

optimumtact replied:
probably because it would be faster to do with a define

MrStonedOne replied:
ya, most likely, but we don't have that either.

optimumtact replied:
can't explain that

MrsTonedOne committed:
debug lines changed to go to admins, debug macros.

MrsTonedOne committed:
Fix bug with mc sleep when set to run less often not calculating ticks correctly

MrsTonedOne committed:
Fixes high pressure movement fuck up

MrsTonedOne committed:
Atmos will zasify round start active turfs to keep them from slowing atmos down

MrStonedOne commented on code/controllers/subsystem/air.dm:
```diff
@@ -272,12 +272,61 @@ var/datum/subsystem/air/SSair
+			active_turfs += new_turfs_to_check
+			turfs_to_check = new_turfs_to_check
+
+		while (turfs_to_check.len)
```
@duncathan I made something to breakdown() all round start active turfs, take a look for me and tell me what you think. The idea being that clearing them prevents them from slowing down atoms the rest of the round. It adds 0.2 to 3.2 (worst case, every turf on the station is round start active) to atmos init.

optimumtact replied:
Can you add a debug define to print the x and y + path of every affected turf? Might help me figure out what the mappers are doing wrong

optimumtact replied:
actually probably better for another PR

MrStonedOne replied:
Debug, debug verbs enable, show at list

optimumtact commented:
Come now mso, quickly!

MrsTonedOne committed:
Cleans things up, final round of fixes and improvements before test merge.

MrStonedOne commented:
This is now test merged on sybil.

optimumtact commented:
@tgstation/commit-access

optimumtact commented on code/__DEFINES/misc.dm:
```diff
@@ -458,3 +457,7 @@ var/global/list/ghost_others_options = list(GHOST_OTHERS_SIMPLE, GHOST_OTHERS_DE
 
+//debug printing macros
+#define debug_world(msg) if (Debug2) world << "DEBUG: [msg]"
+#define debug_admins(msg) if (Debug2) admins << "DEBUG: [msg]"
```
woah does this work?

MrStonedOne replied:
yes, you can use a list in <<.

in fact, i think you can use this for sending client images to avoid the user code level loop, just do list_of_peeps_to_get_client_image << image

you might even be able to do it with a list of images.

optimumtact commented on code/controllers/subsystem/timer.dm:
```diff
@@ -42,38 +52,44 @@ var/datum/subsystem/timer/SStimer
+			return hash_event.id
+	SStimer.hashes[event.hash] = event
+	if (wait <= 0)
+		spawn
```
I'm suprised this is valid syntax and also, what on earth is it doing here? wouldn't it be better to setbackground or something?

MrStonedOne replied:
It runs it immediately if it has a zero wait

optimumtact replied:
ahah, that makes more sense, thanks

Iamgoofball commented on code/controllers/subsystem/air.dm:
```diff
@@ -272,12 +272,61 @@ var/datum/subsystem/air/SSair
+			EG.dismantle()
+
+		var/msg = "HEY! LISTEN! [(world.timeofday - timer)/10] Seconds were wasted processing [starting_ats] world start active turf(s) (connected to [ending_ats] other turfs) with atmos differences at round start."
+		world << "<span class='boldannounce'>[msg]</span>"
```
>world

MrStonedOne commented:
This is ready for re-view, but not quite ready for merge, I got some final cleaning up to do.

Iamgoofball commented on code/controllers/master.dm:
```diff
@@ -43,37 +50,51 @@ var/global/datum/controller/master/Master = new()
+	MC_restart_timeout = world.time + (50 * ++MC_restart_count)
+	Master.processing = -1 //stop ticking this one
+	try
+		new/datum/controller/master()
```
hey what if cleanbot happened with the MC

MrStonedOne replied:
`MC_restart_timeout = world.time + (50 * ++MC_restart_count)` is designed to catch this and throttle the attempts to recreate the mc

Iamgoofball commented on code/controllers/master.dm:
```diff
@@ -120,129 +145,251 @@ var/global/datum/controller/master/Master = new()
+			lobbysubsystems += SS
+		normalsubsystems += SS
+
+	for (var/thing in normalsubsystems|lobbysubsystems)
```
this works????????????????????

RemieRichards replied:
>not knowing basic list operations.

MrsTonedOne committed:
final round of tweaks and bug fixes, first round of cleanups. (still [dwm])

MrStonedOne commented:
todo: some subsystems never run randomly, fix this.

MrsTonedOne committed:
Makes the mc a save a little bit more processing time for byond.

MrsTonedOne committed:
fixes the mc fucking up the first run timing.

MrsTonedOne committed:
trying to find this fucking bug be all like AHHHHHH

MrsTonedOne committed:
Cleans things up, removes debugging lines, adds proper ss level failsafes

MrsTonedOne committed:
some quick debugging lines

MrsTonedOne committed:
KILLING PREFORMANCE TO FIGURE OUT IF THIS IS TRUELY A BYOND BUG

GOING FUCKING INSANE

(also, finishes setting default fps to 20)

MrsTonedOne committed:
FUCK IT, WE'LL DO IT LIVE

I'm done, if the bug still exists, this will massively mitigate the likelihood of it causing issues.

Regardless, i am done looking at this bug, at this code, and at runtimes.

MrStonedOne commented:
Went a whole round on sybil without issues.

I'm calling this ready.

Bawhoppen commented:
god bless america :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us: :us:

HankMoodyExperience commented:
wew lads 20 SILKY SMOOTH FRAMES PER SECOND :+1:

duncathan commented on code/controllers/subsystem/air.dm:
```diff
@@ -274,28 +265,71 @@ var/datum/subsystem/air/SSair
+			EG.add_turf(ET)
+		if (!ET.excited)
+			ET.excited = 1
+			. += ET
```
This is good

Groudonmaster commented:
Very impressive work here stoned
